include/nbl/video/IGPUCommandPool.h

```diff
@@ -31,13 +66,219 @@ class IGPUCommandPool : public core::IReferenceCounted, public IBackendObject
         // Vulkan: const VkCommandPool*
         virtual const void* getNativeHandle() const = 0;
 
+        core::CMemoryPool<core::GeneralpurposeAddressAllocator<uint32_t>, core::default_aligned_allocator, false, uint32_t> m_commandSegmentPool;
+
     protected:
         virtual ~IGPUCommandPool() = default;
 
         core::bitflag<E_CREATE_FLAGS> m_flags;
         uint32_t m_familyIx;
 };
 
+class alignas(COMMAND_ALIGNMENT) IGPUCommandPool::ICommand
+{
+public:
+    virtual ~ICommand() {}
+
+    // static void* operator new(std::size_t size) = delete;
+    static void* operator new[](std::size_t size) = delete;
+    // static void* operator new(std::size_t size, std::align_val_t al) = delete;
+    static void* operator new[](std::size_t size, std::align_val_t al) = delete;
+
+    // static void operator delete  (void* ptr) = delete;
+    static void operator delete[](void* ptr) = delete;
+    static void operator delete  (void* ptr, std::align_val_t al) = delete;
+    static void operator delete[](void* ptr, std::align_val_t al) = delete;
+    static void operator delete  (void* ptr, std::size_t sz) = delete;
+    static void operator delete[](void* ptr, std::size_t sz) = delete;
+    static void operator delete  (void* ptr, std::size_t sz, std::align_val_t al) = delete;
+    static void operator delete[](void* ptr, std::size_t sz, std::align_val_t al) = delete;
+
+    uint32_t m_size;
+protected:
+    ICommand(uint32_t size) : m_size(size) {}
+};
+
+class alignas(COMMAND_SEGMENT_ALIGNMENT) IGPUCommandPool::CommandSegment
+{
+public:
+    struct Iterator;
+
+    struct params_t
+    {
+        core::LinearAddressAllocator<uint32_t> m_commandAllocator;
+        CommandSegment* m_next;
+    } params;
+
+    static inline constexpr uint32_t STORAGE_SIZE = COMMAND_SEGMENT_SIZE - sizeof(params_t);
+    alignas(ICommand) uint8_t m_data[STORAGE_SIZE];
+
+    CommandSegment()
+    {
+        params.m_commandAllocator = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, COMMAND_SEGMENT_ALIGNMENT, COMMAND_SEGMENT_SIZE);
+        params.m_next = nullptr;
+
+        wipeNextCommandSize();
+    }
+
+    template <typename Cmd, typename... Args>
+    // Cmd* allocate(const Args&... args)
+    Cmd* allocate(Args&&... args)
```

Comment:
yes `&&` you want perfect forwarding

Reply:
also if you're doing to be doing both allocation and construction, then you better call it `emplace`

Reply:
actually its a bad idea to have the construction and fowarding here, see my big PR message.

make it a `Cmd* allocate(const Args&... args)` and **do not construct the command in here**

Reply:
Resolved.